esphome/components/ds3231/ds3231.h

```diff
@@ -0,0 +1,215 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/components/i2c/i2c.h"
+#include "esphome/components/sensor/sensor.h"
+#include "esphome/components/time/real_time_clock.h"
+
+namespace esphome {
+namespace ds3231 {
+
+enum DS3231Alarm1Mode {
+  EVERY_SECOND = 0x0F,
+  MATCH_SECOND = 0x0E,
+  MATCH_MINUTE_SECOND = 0x0C,
+  MATCH_HOUR_MINUTE_SECOND = 0x08,
+  MATCH_DAY_OF_MONTH_HOUR_MINUTE_SECOND = 0x00,
+  MATCH_DAY_OF_WEEK_HOUR_MINUTE_SECOND = 0x10,
+};
+
+enum DS3231Alarm2Mode {
+  EVERY_MINUTE = 0x0E,
+  MATCH_MINUTE = 0x0C,
+  MATCH_HOUR_MINUTE = 0x08,
+  MATCH_DAY_OF_MONTH_HOUR_MINUTE = 0x00,
+  MATCH_DAY_OF_WEEK_HOUR_MINUTE = 0x10,
+};
+
+enum DS3231SquareWaveMode {
+  MODE_SQUARE_WAVE,
+  MODE_ALARM_INTERRUPT,
+};
+
+enum DS3231SquareWaveFrequency {
+  FREQUENCY_1_HZ = 0x00,
+  FREQUENCY_1024_HZ = 0x01,
+  FREQUENCY_4096_HZ = 0x02,
+  FREQUENCY_8192_HZ = 0x03,
+};
+
+class DS3231RTC;
+
+class DS3231Sensor;
+
+class DS3231Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void set_default_alarm_1(DS3231Alarm1Mode mode, bool int_enabled, uint8_t second, uint8_t minute, uint8_t hour,
+                           uint8_t day);
+  void set_default_alarm_2(DS3231Alarm2Mode mode, bool int_enabled, uint8_t minute, uint8_t hour, uint8_t day);
+  void set_default_square_wave_mode(DS3231SquareWaveMode mode) { this->square_wave_mode_ = mode; }
+  void set_default_square_wave_frequency(DS3231SquareWaveFrequency frequency) {
+    this->square_wave_frequency_ = frequency;
+  }
+  void add_on_alarm_callback(std::function<void(uint8_t)> &&callback);
+  float get_setup_priority() const override { return setup_priority::DATA; }
```

Comment:
I think HARDWARE makes more sense here, it's not as if this component depends on the data from others.
```suggestion
  float get_setup_priority() const override { return setup_priority::HARDWARE; }
```